File: 39-combination-sum/39-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */

// backtracking decision tree problem
const combinationSum = function(candidates, target) {
    const res = [];
    
    const dfs = (comb, sum, index) => {
        if (sum > target) return;
        if (sum === target) {
            res.push([...comb]);
            return;
        }
         
        for (let i = index; i < candidates.length; i++) {
            // choose
            comb.push(candidates[i]);
            // explore
            dfs(comb, sum + candidates[i], i);
            // unchoose
            comb.pop();
        }
    }
    
    dfs([], 0, 0);
    
    return res;
};

File: 39-combination-sum/39-combination-sum.rb
# @param {Integer[]} candidates
# @param {Integer} target
# @return {Integer[][]}
def combination_sum(candidates, target)
    res = []
    dfs(0, 0, candidates, target, res, [])
    res
end

def dfs(index, sum, nums, target, res, comb)
    return res << comb if sum == target
    return if sum > target || index == nums.size
    
    (index...nums.size).each do |i|
        dfs(i, sum + nums[i], nums, target, res, [*comb, nums[i]]) 
    end
end